### PR TITLE
Crossfilter bug fixes and documentation

### LINKED ISSUE #1703: Extending/Fixing Crossfilter
I was looking at extending the crossfilter functionality, since I've previously built something similar in Matlab.

When working on the current master branch, I noticed that not everything is working how it should. First I started walking through some of the functions under models and added documentation as I understood more about what they were doing. Then, I started fixing some issues that appear to exist due to some change in the interface for the main bokeh functionality.

I wanted to go ahead and reference [this commit](https://github.com/rothnic/bokeh/commit/e338c26adf9b9456f8cd68c43cd66208881c7464) to see if there is someone already working in this area, and see if extending it would be of interest.

### Use Cases
I think there are two major use cases that might be different apps, but overlap somewhat:
* Tableau-like Dashboard creation (create a plot, create another, piece them together with controls)
* Exploratory tool that provides faceting (seems somewhat incompatible with the first use case)

### Features
* More plot types
* More aesthetics to map columns into (color, size, shape, etc)
* "Saving" plots to use in a larger dashboard
* Gracefully handle more columns (big list to scroll through, maybe collapse summary by default)
* Handling multiple data sets (configured to different plots on a dashboard, or combining through simple joins of some kind)

### PR DESCRIPTION
**Pull request by rothnic**
Issues: fixes #1703

Going ahead and opening this pull request for discussion. It is nearly complete, containing a lot of documentation additions, and fixes a few bugs. Might be a little overboard with comments/docs, but good for an example application.

I left some todos in there that I'll clean up before finalizing it and there is one remaining bug I've found that I cannot track down. Result is an error on the javascript side when creating a scatter plot with categorical data on the x axis and continuous on the y axis. The result is an empty plot area.

### CODE REVIEW

**Comment by damianavila** [comment]
@rothnic, at the top of the PR description I added a reference to the issue discussing some of the things you implemented here... we usually link **Issues** with **PRs** (you can know more about our Issues and PR management guidelines here: [BEP 1](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management)). Since you are contributing regularly, and I hope this continue :wink: for a lot of time, I am just letting you know some of the procedures we follow in the daily work inside Bokeh.
OK, coming back to the implementation itself, I will take a look soon (Friday night here, so time to go out for a :beer:)

**Comment by rothnic** [comment]
Remaining issues are:
* Failed FactorRange Test: due to the FactorRange modification. Need to look into which side needs to be fixed. Crossfilter or FactorRange and the test.
* Broken Scatter Plot when one dimension is mapped to a discrete source and one a continuous. See [this comment](https://github.com/bokeh/bokeh/issues/1703#issuecomment-70529654) for more detail.

File: examples/app/crossfilter/crossfilter_app.py [hunk 1]
```diff
@@ -9,6 +9,6 @@
 @object_page("crossfilter")
 def make_object():
     autompg['cyl'] = autompg['cyl'].astype(str)
-    autompg['origin'] = autompg['cyl'].astype(str)
```
Comment (rothnic) [inline examples/app/crossfilter/crossfilter_app.py]: I believe this was just a copy/paste error. This made origin have the same data as cyl column.

Comment (bryevdv) [inline examples/app/crossfilter/crossfilter_app.py]: Yup, looks like

**Comment by rothnic** [comment]
Left a bunch of comments on the latest commit, which should fix the test failure. This also fixes the remaining bugs that I identified, and provides some communication back if it is configured in a "bad" way for more graceful failure.

Pending any issues noticed, I think this is ready.

**Comment by bryevdv** [comment]
@rothnic this is all brilliant. I will plan to try and get this merged fairly early tomorrow. Thanks for all the good work!

**Comment by bryevdv** [comment]
Checking it out now, hope to merge in the next hour or so.

**Comment by bryevdv** [comment]
@rothnic things are definitely more robust, thanks for this contribution. Do you intend to continue working on improving the crossfilter in additional PRs? If so we should create a "master" issue to collect all the potential improvements that could be done as sub-issues.

File: bokeh/crossfilter/models.py [hunk 2]
```diff
@@ -128,84 +219,174 @@ def set_input_selector(self):
...
+        # no faceting
+        if all([len(self.facet_x) == 0,
                 len(self.facet_y) == 0,
                 len(self.facet_tab) == 0]):
             return self.make_single_plot()
 
-        if all([len(self.facet_x) !=0,
-                len(self.facet_y) == 0,
+        # x xor y faceting
+        if all([(len(self.facet_x) != 0) ^ (len(self.facet_y) != 0),
```
Comment (damianavila) [inline bokeh/crossfilter/models.py]: I would probably use `if all([(len(self.facet_x) != 0) != (len(self.facet_y) != 0),` but the bitwise is valid too...

File: bokeh/crossfilter/models.py [hunk 3]
```diff
@@ -238,83 +439,208 @@ def make_xy_facet_plot(self):
...
+        # return the grid of plots as the plot
+        grid = GridPlot(children=grid_plots, plot_width=200*len(all_facets_x))
+        return grid
 
     def make_single_plot(self, df=None, title=None,
                          plot_width=700, plot_height=680,
-                         tools="pan,wheel_zoom,box_zoom,save,resize,select,reset"
-                     ):
+                         tools="pan,wheel_zoom,box_zoom,save,resize,"
+                               "box_select,reset"):
```
Comment (damianavila) [inline bokeh/crossfilter/models.py]: You can leave all in one line here. For readability, I think it is better so have one line, even if we are above the 80 chars...

**Pull request merged**